### DISCUSSION TITLE
How to serve static or .php files

### DISCUSSION THREAD

**Discussion by rabol** [laravel/valet#1430] [Q&A] [answered]
Hi

Sometimes I have a need to serve files from a subfolder of a Valet site, could be a static html or a .php file.
My normal setup is for Laravel development, but if Valet is using nginx in the background it should be possible to serve static html or .php files

How do I configure Valet to do the above ?

Thanks

**Comment by drbyte** [comment]
There's some built-in detection for identifying static assets and serving them direction, but it's more geared around images and stylesheets.

You could implement a custom driver that detects certain attributes about your non-Laravel project and therefore loads itself instead of the Laravel driver, and then your custom functionality could serve those files with more specificity.

Sorry if that sounds "vague". 

Can you share a few examples? (Sure, I could guess at some potentials, but some specifics from your use-case would make it easier to set something up to explore.)
And, for the .php file scenario, are you talking about a standalone PHP file, or something that also loads a bunch of other files too?

**Comment by rabol** [comment]
One use case I had recently was that I needed evaluate a paid 3rd party library and they required Ioncube to be installed, the recommended way to install Ioncube is to copy a installation folder to the root of the web and then execute the installation script - that does not work.

Another was to get PhpStorm working with XDebug I needed the phpinfo() result so I created a simple file

```
<?php
phpinfo();
```
and put it in my public folder, that does not work

on my current project I have Terms and condition, privacy policy pages in different languages, and these pages does not change very often and have more or less no need for php so I wanted them as static html, but that does not work either 


Sure, I found workaround for all, but...

**Comment by drbyte** [comment, marked as answer]
Here you go...

Create a new custom driver file at: `/Users/YOUR_USERNAME/.config/valet/Drivers/StaticLaravelValetDriver.php` containing the following:

```php
<?php
namespace Valet\Drivers\Custom;

use Valet\Drivers\LaravelValetDriver;

class StaticLaravelValetDriver extends LaravelValetDriver
{
    public function frontControllerPath(string $sitePath, string $siteName, string $uri): ?string
    {
        if (file_exists($staticFilePath = $sitePath.'/public'.$uri)
           && $this->isActualFile($staticFilePath)) {
            return $staticFilePath;
        }

        return $sitePath.'/public/index.php';
    }
}
```

This will cause this custom override driver to be used with your Laravel sites, and when `/public/*.*` static files are found (including .php files), they will be served before calling Laravel via `public/index.php`

Why is this needed? Because (with the present LaravelValetDriver) Valet doesn't check for static `.php` files in the public directory, since if it were to serve them via its static-file-handling functions it would recursively load itself endlessly. Way back in an early version of Valet the static-file-handling was added to be more performant and to cover broader use-cases, but the .php support for static files was dropped for safety. But you can certainly use your own custom driver (as above) to accommodate this.
